Decorator should honour implementation checks

krmahadevan opened the pull request:
Fixes #11432

The current web driver decorator looks for 
Class name compatibility against a bunch of interfaces.

So this works with we use WebDriver but does not
Work when we use RemoteWebDriver.

Altered the logic to use assignability checks between classes instead of doing a equals check.

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->
The current web driver decorator looks for 
Class name compatibility against a bunch of interfaces.

So this works with we use WebDriver but does not
Work when we use RemoteWebDriver.

Altered the logic to use assignability checks between classes instead of doing a equals check.
### Description
<!--- Describe your changes in detail -->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

pujagani commented on java/test/org/openqa/selenium/support/events/EventFiringDecoratorTest.java:
```diff
@@ -436,4 +438,21 @@ public void onError(Object target, Method method, Object[] args, InvocationTarge
   }
+
+  @Test
+  public void ensureListenersAreInvoked() {
```
Probably rename it to depict that decorating the Webdriver subclass works.

pujagani commented on java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java:
```diff
@@ -304,13 +304,20 @@ protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<Z> clazz)
           return method.invoke(decorated, args);
         }
-        if (originalInterfaces.contains(method.getDeclaringClass())) {
+        boolean isCompatible = originalInterfaces.stream()
```
I understand the checks that are done to solve the issue. But it might be a good idea to name them separately instead of "isCompatible". It might help anyone reading the code understand what each does. Just a suggestion.

krmahadevan replied:
@pujagani - I wasn't able to come up with a good descriptive name (which is short enough as well and not overly verbose) which explains very clearly what is happening. So I have instead added some comments that should help explain what is going on. Happy to rename the variable to something else if you can suggest something which you feel is more apt. We are essentially doing a compatibility check which is why I chose to name it as `isCompatible`

krmahadevan replied:
@pujagani - I wasn't able to come up with a good descriptive name (which is short enough as well and not overly verbose) which explains very clearly what is happening. So I have instead added some comments that should help explain what is going on. Happy to rename the variable to something else if you can suggest something which you feel is more apt. We are essentially doing a compatibility check which is why I chose to name it as `isCompatible`

pujagani replied:
I think the comments make the intent clear now. Thank you!

pujagani commented:
I debugged through the fix and it is solving the issue accurately. Overall, PR looks good to me. Once the comments are addressed and all checks pass, I will merge the PR. Thank you!

krmahadevan commented:
> I debugged through the fix and it is solving the issue accurately. Overall, PR looks good to me. Once the comments are addressed and all checks pass, I will merge the PR. Thank you!

I have addressed the review comment and responded to the other one. Also rebased off of trunk and force pushed the amended commit. Please help take a look

pujagani reviewed (approved):
Thank you @krmahadevan!
